torch/csrc/jit/codegen/cuda/root_domain_map.cpp

```diff
@@ -1095,9 +1091,10 @@ bool ComputeAtRootDomainMapBuilder::safeToMap(const DomainKeySet& domains) {
       unique_domains.push_back(domain);
     }
   }
-  if (hasMatchingDomains(unique_domains)) {
-    return false;
-  }
+  // if (hasMatchingDomains(unique_domains)) {
+  // Fails in view operations, check FusionViewMagicSchedule3
+  //   return false;
+  // }
```

Comment:
This seems just like technical debt from a world that isn't considering view. Specifically:

```.cpp
// Checks whether all consumers of a producer can be joined without
// introducing unsupported mappings. Specifically, if a domain of a
// consumer has a mapped iteration domain in another consumer that
// does not correspond to the same producer iteration domain, mapping
// the consumer domains would result in the producer iteration domain
// mapped to two different consumer iteration domains, requiring
// recomputations.
bool ComputeAtRootDomainMapBuilder::hasMatchingDomains(
```
Seems to conflict with:
```.cpp
std::unordered_set<IterDomain*> ComputeAtRootDomainMap::getMappableDims(
    const TensorDomain* producer,
    const TensorDomain* consumer) const {
  //! This funciton previously used mapBestEffort but it can fail when
  //! a domain is mapped to multitple domains, which can happen with
  //! views. Since we only need to find mappable domains, just
  //! grab any domain that is mapped in a pairwise way.
```

Reply:
Removed in #1941

Reply:
Yeah, it's not supported. I added a check in #1941 that should flag patterns like the one you mentioned.

Reply:
Should we prioritize this issue?

Reply:
That issue is not a blocker for me, so I am fine with it. But seems that with you PR, solving that issue will be as simple as adding something like
```c++
if (failIfSelfMappingExists(fusion, idgraph)) {
  return false;
}
```
? If so, then probably we should do that issue.

Reply:
It doesn't solve the issue as `failIfSelfMappingExists` just throws an exception with `TORCH_INTERNA_ASSERT`. Or, maybe I'm misunderstanding what you are saying, but in any case, we would need to revisit many parts of the lowering to support such a pattern.

Reply:
Can we make self-mapping not a hard failure, but just add a new method `id_graph.hasSelfMapping()`? If so, we can then in this function
https://github.com/csarofeen/pytorch/blob/631094891a96f715d8c9925fb73d41013ca7f2e3/torch/csrc/jit/codegen/cuda/scheduler/registry.cpp#L1304-L1325
We can just add a check
```c++
if (id_graph.hasSelfMapping()) {
  return false;
}
```
And this will causes the segmenter to segment the fusion into pieces that does not have self mapping.

Reply:
And we don't need to do anything in the lowering if we always segment fusions into pieces without self mapping.

Reply:
That sounds good. Probably we should this discussion to a different issue or PR as it's not directly related to this PR.

Reply:
Sure